### PR TITLE
treefmt 0.4.1 (new formula)

### PR DESCRIPTION
**Pull request by mloughran**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Adds [treefmt](https://github.com/numtide/treefmt).

### CODE REVIEW

File: Formula/treefmt.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  depends_on "rust" => :build
+
+  def install
+    system "cargo", "install", "--path", ".", "--root", buildpath
+    bin.install buildpath / "bin" / "treefmt"
+  end
+
+  test do
+    system bin / "treefmt", "-V"
```
Comment (chenrui333) [inline Formula/treefmt.rb:18]: Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

File: Formula/treefmt.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  sha256 "48554e1b030781c49c98c5882369a92e475d76fee0d5ce2d2f79966826447086"
+  license "MIT"
+
+  head "https://github.com/numtide/treefmt.git", branch: "master"
+
+  depends_on "rust" => :build
+
+  def install
+    system "cargo", "install", "--path", ".", "--root", buildpath
+    bin.install buildpath / "bin" / "treefmt"
```
Comment (carlocab) [inline Formula/treefmt.rb:14]:
```suggestion
    system "cargo", "install", *std_cargo_args
```
Does this not work?

Comment (mloughran) [reply Formula/treefmt.rb:14]: @carlocab Thanks, I hadn't seen `std_cargo_args` documented. It seems to work but these additional crates files are installed. I assume this is not desirable?

```
% brew list treefmt
/opt/homebrew/Cellar/treefmt/0.4.1/.crates.toml
/opt/homebrew/Cellar/treefmt/0.4.1/.crates2.json
/opt/homebrew/Cellar/treefmt/0.4.1/bin/treefm
```

Comment (carlocab) [reply Formula/treefmt.rb:14]: That's fine. You can remove them if you like, but I don't think removing them is worth the extra code. (Avoiding them isn't worth not using `std_cargo_args` either.)

Comment (mloughran) [reply Formula/treefmt.rb]: It doesn't bother me either way so have made accepted the change.

File: Formula/treefmt.rb (hunk 1 above)
Comment (mloughran) [reply Formula/treefmt.rb:17]: Testing would require also installing a language formatter, including some example code to be formatted, and running treefmt. Personally I don't think this is worthwhile given the version check demonstrates that compilation was successful, which is the only thing this formula does. @zimbatm can you think of a simple way of testing?

Comment (SMillerDev) [reply Formula/treefmt.rb:17]: > Personally I don't think this is worthwhile given the version check demonstrates that compilation was successful

Given we have a couple years of experience packaging software, you can trust me when I say a successful compilation does not ensure that your program actually works for users.

Comment (carlocab) [reply Formula/treefmt.rb:17]: You can also try running `treefmt` with a missing language formatter and check that we get the expected error.

Comment (zimbatm) [reply Formula/treefmt.rb:17]: I propose a fake formatter that tests a few of treefmt's behaviours:

`treefmt.toml`
```toml
[formatter.test]
includes = [ "*.test" ]
command = "/bin/sh"
options = [
  "-ec",
  """
  for f in "$@"; do
    case "$f" in
    *nochange.test)
      : # nothing to do, is formatted
      ;;
    *reformat.test)
      echo "needed reformatting" > "$f"
      ;;
    *)
      echo "FAIL: unknown file"
      exit 1
      ;;
    esac
  done
  """
]
```
Then a test could look like this in bash:
```
set -euo pipefail

# Some fixture
touch nochange.test reformat.test
nochange_stat=$(stat nochange.test)
reformat_stat=$(state reformat.test)

# Run the formatter, it should exit with success
treefmt

# This file should not have changed
[[ "$nochange_stat" = "$(stat nochange.test)" ]]
# This file should have changed
[[ "$reformat_stat" != "$(stat reformat.test)" ]]
[[ $(< reformat.test) = "needed reformatting" ]]
```

Comment (mloughran) [reply Formula/treefmt.rb]: > You can also try running treefmt with a missing language formatter and check that we get the expected error.

@carlocab done in https://github.com/Homebrew/homebrew-core/pull/105512/commits/16c81121d858ccf13a946051b238ab2fb51ed2f4

@zimbatm thanks for the suggestion, however unless there is strong pushback my view is that this is not the correct place to be testing treefmt behaviour. It's already a lot of work to implement and keep updated definitions for each package management system (as you well know!), so I think a pragmatic balance needs to be struck between thoroughness and simplicity.

**Review by carlocab — APPROVED** [review]
Thanks @mloughran!

**Pull request closed without merging**